core/src/main/java/org/elasticsearch/indices/cluster/IndicesClusterStateService.java

```diff
@@ -148,12 +148,17 @@ public IndicesClusterStateService(Settings settings, IndicesService indicesServi
 
     @Override
     protected void doStart() {
-        clusterService.addFirst(this);
+        // Doesn't make sense to manage shards on non-master and non-data nodes
+        if (DiscoveryNode.isDataNode(settings) || DiscoveryNode.isMasterNode(settings)) {
```

Comment:
ClusterService has not been started at that point and `clusterService.localNode()` throws an `java.lang.IllegalStateException` when invoked here. We might do some reordering in the initialization logic etc. but as this commit targets multiple versions I wanted the change to be simple.